### ISSUE TITLE
Type instantiation is excessively deep and possibly infinite.

### ISSUE THREAD

**Issue by jasonkuhrt** [vitest-dev/vitest#6546]
### Describe the bug

There is an issue in the `expectTypeOf` type utility. This may be an upstream issue. If so sorry in advance.

Notice I rolled my own type equality check that passes.

![CleanShot 2024-09-22 at 09 54 29@2x](https://github.com/user-attachments/assets/c49b5735-d198-4a0e-aee7-48b3b4fd1b82)

![CleanShot 2024-09-22 at 09 54 55@2x](https://github.com/user-attachments/assets/9c0dc6d8-7d32-4361-a44d-3f01ba365a71)

### Reproduction

Available at this commit [jasonkuhrt/graffle@`eeb4ba4` (#1107)](https://github.com/jasonkuhrt/graffle/pull/1107/commits/eeb4ba4263df020a344e0ded64414887cab59b22) at /src/layers/4_ResultSet/types.test-d.ts

### System Info

```shell
System:
    OS: macOS 15.0
    CPU: (8) arm64 Apple M1 Pro
    Memory: 633.48 MB / 32.00 GB
    Shell: 3.5.0 - /opt/homebrew/bin/fish
  Binaries:
    Node: 22.7.0 - ~/Library/pnpm/node
    npm: 10.8.2 - ~/Library/pnpm/npm
    pnpm: 9.10.0 - /opt/homebrew/bin/pnpm
  Browsers:
    Brave Browser: 114.1.52.122
    Chrome: 128.0.6613.139
    Safari: 18.0
  npmPackages:
    vitest: ^2.1.1 => 2.1.1
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by jasonkuhrt** [comment]
Right, maybe its an issue of https://github.com/mmkal/expect-type.

**Comment by aryaemami59** [comment]
@jasonkuhrt Do you happen to have an actual reproduction? I can't seem to replicate it.

**Comment by jasonkuhrt** [comment]
Hey @aryaemami59, once I wrap this this PR https://github.com/jasonkuhrt/graffle/pull/1107 I'll try to get a more minimal repro going. Thanks for your work on this!

**Comment by sheremet-va** [comment]
Doesn't seem like this can be fixed by Vitest.
